lib/middleware/handle_errors.ex

```diff
@@ -53,4 +62,15 @@ defmodule DeliriumTremex.Middleware.HandleErrors do
       messages: ["Something went wrong"]
     }
   end
+
+  defp is_changeset?(error) do
+    case Code.ensure_compiled(Ecto.Changeset) do
+      {:module, Ecto.Changeset} -> !Map.get(error, :valid?)
```

Comment:
Shouldn't this return true or false no matter if the changeset is valid or not?
It doesn't make sense for this function to fail if the changeset is valid - it;s unexpected given the name and context of the function.

```suggestion
      {:module, Ecto.Changeset} -> true
```

Reply:
I commited this because I wanted some sort of check if the error in that case truly is an `%Ecto.Changeset{}` struct. I also thought about adding something like 
`Map.has_key?(error, :__struct__)`.
If you think that's unnecessary I'm fine with removing it.

Reply:
> I commited this because I wanted some sort of check if the error in that case truly is an `%Ecto.Changeset{}` struct.

I don't get it, you checked that with `Code.ensure_compiled(Ecto.Changeset)` and the pattern match. From what I understand you are checking if the changeset is valid or not with `!Map.get(error, :valid?)` - which, in my opinion, is confusing to the reader because the function `is_changeset?` communicates that it checks if something is a changeset.

>  I also thought about adding something like `Map.has_key?(error, :__struct__)`

I think that `Code.ensure_compiled(Ecto.Changeset)` is good enough for this purpose.

> If you think that's unnecessary I'm fine with removing it.

If your goal was to check if this is an Ecto changeset, then I would prefer for this to be removed. If you had other goals I would like to discuss them.

Reply:
Yes, the goal was to check if the error is an Ecto changeset. As far as I understand we handle four possible errors: an atom, a map, an Ecto changeset and an unknown error in case that any of these don't match.

How will we prevent an unhandled error to enter the `DeliriumTremex.Formatters.Ecto.Changeset` module if we remove this check and don't implement something else? I don't see how `Code.ensure_compiled(Ecto.Changeset)` is enough.

Reply:
Aha. I get it now. The I would suggestim something like

```suggestion
      {:module, Ecto.Changeset} -> 
        case error do
          %Ecto.Changeset{} -> true
          _ -> false
        end
```

Reply:
Actually, a `with` would be much better than a case here

```elixir
with {:module, _} <- Code.ensure_compiled(Ecto.Changeset),
  %Ecto.Changeset{} <- error do
  true
else
  false
end
```

Reply:
I don't think that will compile since we don't have Ecto anymore in our dependencies and `%Ecto.Changeset{}` struct is undefined.

Reply:
As far as I know structs are resolved at runtime therefore it should work.
Just as an example, you are matching against `{:module, Ecto.Changeset}` which also has Ecto in it, and it works.

Reply:
That's a module name not a definition of a struct. I've just tried it and:
```
== Compilation error in file lib/middleware/handle_errors.ex ==
** (CompileError) lib/middleware/handle_errors.ex:70: Ecto.Changeset.__struct__/0 is undefined, cannot expand struct Ecto.Changeset
    (stdlib) lists.erl:1354: :lists.mapfoldl/3
```

Reply:
Perhaps then `error.__struct__ == Ecto.Changeset` ?

Reply:
Great, I like that one! :+1: 

This is my suggestion:
```
  defp is_changeset?(error) do
    with \
      {:module, Ecto.Changeset} <- Code.ensure_compiled(Ecto.Changeset),
      Ecto.Changeset <- Map.get(error, :__struct__)
    do
      true
    else
      _ -> false
    end
  end
```